src/connection.rs

```diff
@@ -46,29 +46,44 @@ impl Connection {
         Ok(self.transformer.deserialize()?.0)
     }
 
-    async fn read_and_deserialize<'a, T>(&'a mut self, size: usize) -> Result<T>
-    where
-        T: Deserialize<'a>,
-    {
-        self.transformer.get_buffer().clear();
-        read_exact_to_vec(
-            &mut self.raw_conn.stream,
-            self.transformer.get_buffer(),
-            size,
-        )
-        .await?;
+    pub(crate) async fn read_response(&mut self) -> Result<Response> {
+        // Clear content of buffer
+        //
+        // We assume that previous call to `Connection::read_response` does not
+        // read any bytes of the current packet to read in.
+        let buffer = self.transformer.get_buffer();
+        buffer.clear();
 
-        self.deserialize()
-    }
+        // Deserialize len of the packet (first 4b)
+        //
+        // We will read in at most 16 bytes here to cache the whole packet
+        // because any response takes at least 16 bytes
+        //  - 0..4 len of packet
+        //  - 4..8 type of the packet
+        //  - 8..16 version (`Response::Hello`), `response_id` or `session_id`.
+        //
+        // Reading more than that will break precondition of next call to
+        // `Connection::read_response`.
+        read_to_vec_rng(&mut self.raw_conn.stream, buffer, 4..16).await?;
 
-    /// Return size of the response.
-    async fn read_header(&mut self) -> Result<u32> {
-        self.read_and_deserialize(4).await
-    }
+        let len: u32 = self.deserialize()?;
 
-    pub(crate) async fn read_response(&mut self) -> Result<Response> {
-        let len = self.read_header().await?;
-        self.read_and_deserialize(len as usize).await
+        // Remove first 4b representing the len
+        let buffer = self.transformer.get_buffer();
+        buffer.drain(..4);
+
+        // Read in rest of the packet
+        let len: usize = len.try_into().unwrap();
+
+        // Make sure we won't accidentally read in the next packet
+        // since `Connection::read_response` clear the buffer before
+        // reading anything.
+        debug_assert!(buffer.len() <= len);
```

Comment:
I'm not sure I follow why we do it this way? Wouldn't it be better to not clear the buffer on each pass, and instead just consume `len` bytes from `buffer` and leave whatever trails it behind in the buffer for the next call?

Also, why is it okay for `buffer.len() < len`?

Reply:
Because the same buffer is used for serializing.

> Also, why is it okay for buffer.len() < len?

The read_in_vec_rng reads at most 16 bytes, which is smaller than the maximum size of a packet.

Reply:
Done!